File: backend/src/services/embeddings.js
import { GoogleGenerativeAI } from '@google/generative-ai';
import { config } from '../config/index.js';

const genAI = new GoogleGenerativeAI(config.geminiApiKey);

// Embeds a single string or batch of strings. Returns a single vector for a
// string input, or an array of vectors for an array input.
export async function embedText(textOrArray) {
  const model = genAI.getGenerativeModel({ model: config.embeddingModel });

  if (Array.isArray(textOrArray)) {
    const { embeddings } = await model.batchEmbedContents({
      requests: textOrArray.map((text) => ({
        content: { role: 'user', parts: [{ text }] },
        outputDimensionality: config.embeddingDimensions,
      })),
    });
    return embeddings.map((e) => e.values);
  }

  const { embedding } = await model.embedContent({
    content: { role: 'user', parts: [{ text: textOrArray }] },
    outputDimensionality: config.embeddingDimensions,
  });
  return embedding.values;
}

File: backend-py/app/services/embeddings.py
"""Gemini embeddings via the google-genai SDK (async)."""
from __future__ import annotations

from google.genai import types

from ..config import settings
from .gemini_client import client, gemini_retry

# Gemini's embedding API accepts batches; keep them modest to stay within limits.
_BATCH_SIZE = 100


@gemini_retry
async def _embed_batch(batch: list[str], task_type: str):
    return await client.aio.models.embed_content(
        model=settings.embedding_model,
        contents=batch,
        config=types.EmbedContentConfig(
            output_dimensionality=settings.embedding_dimensions,
            task_type=task_type,
        ),
    )


async def embed_texts(texts: list[str]) -> list[list[float]]:
    """Embed a list of strings, returning one 768-dim vector per input (order preserved)."""
    if not texts:
        return []

    vectors: list[list[float]] = []
    for start in range(0, len(texts), _BATCH_SIZE):
        batch = texts[start : start + _BATCH_SIZE]
        result = await _embed_batch(batch, "RETRIEVAL_DOCUMENT")
        vectors.extend([e.values for e in result.embeddings])
    return vectors


async def embed_query(text: str) -> list[float]:
    """Embed a single query string (uses the RETRIEVAL_QUERY task type)."""
    result = await _embed_batch([text], "RETRIEVAL_QUERY")
    return result.embeddings[0].values
